// angular-date-value-accessor/esm2015/lib/date-value-accessor.js
import { Directive, ElementRef, HostListener, Renderer2, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export const DATE_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => DateValueAccessor),
    multi: true
};
/**
 * The accessor for writing a value and listening to changes on a date input element
 *
 *  ### Example
 *  `<input type="date" name="myBirthday" ngModel useValueAsDate>`
 */
// tslint:disable-next-line: directive-class-suffix
export class DateValueAccessor {
    constructor(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.onChange = (_) => { };
        this.onTouched = () => { };
    }
    writeValue(value) {
        if (!value) {
            this.renderer.setProperty(this.elementRef.nativeElement, 'value', null);
            return;
        }
        this.renderer.setProperty(this.elementRef.nativeElement, 'valueAsDate', value);
    }
    registerOnChange(fn) { this.onChange = fn; }
    registerOnTouched(fn) { this.onTouched = fn; }
    setDisabledState(isDisabled) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'disabled', isDisabled);
    }
}
DateValueAccessor.decorators = [
    { type: Directive, args: [{
                // this selector changes the previous behavior silently and might break existing code
                // selector: 'input[type=date][formControlName],input[type=date][formControl],input[type=date][ngModel]',
                // this selector is an opt-in version
                // tslint:disable-next-line: directive-selector
                selector: '[useValueAsDate]',
                providers: [DATE_VALUE_ACCESSOR]
            },] }
];
DateValueAccessor.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
DateValueAccessor.propDecorators = {
    onChange: [{ type: HostListener, args: ['input', ['$event.target.valueAsDate'],] }],
    onTouched: [{ type: HostListener, args: ['blur', [],] }]
};
