// angular-date-value-accessor/esm2015/lib/local-date-value-accessor.js
import { Directive, ElementRef, HostListener, Renderer2, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export const LOCAL_DATE_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => LocalDateValueAccessor),
    multi: true
};
/**
 * The accessor for writing a value and listening to changes on a date input element
 *
 *  ### Example
 *  `<input type="date" name="myBirthday" ngModel useValueAsLocalDate>`
 */
// tslint:disable-next-line: directive-class-suffix
export class LocalDateValueAccessor {
    constructor(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.onChange = () => { };
        this.onInput = (date) => {
            let selectedDate = null;
            if (date) {
                // Create LOCAL Date, time is set to 00:00 in LOCAL time
                selectedDate = new Date(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate());
                ;
            }
            this.onChange(selectedDate);
        };
        this.onTouched = () => { };
    }
    writeValue(date) {
        // Create UTC Date, time is set to 00:00 in UTC time
        const utcDate = date ?
            new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate())) :
            null;
        this.renderer.setProperty(this.elementRef.nativeElement, 'valueAsDate', utcDate);
    }
    registerOnChange(fn) { this.onChange = fn; }
    registerOnTouched(fn) { this.onTouched = fn; }
    setDisabledState(isDisabled) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'disabled', isDisabled);
    }
}
LocalDateValueAccessor.decorators = [
    { type: Directive, args: [{
                selector: '[useValueAsLocalDate]',
                providers: [LOCAL_DATE_VALUE_ACCESSOR]
            },] }
];
LocalDateValueAccessor.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
LocalDateValueAccessor.propDecorators = {
    onInput: [{ type: HostListener, args: ['input', ['$event.target.valueAsDate'],] }],
    onTouched: [{ type: HostListener, args: ['blur', [],] }]
};
// Use Local Dates with html input type date
// https://stackoverflow.com/questions/53032953/what-is-the-correct-way-to-set-and-get-htmlinputelement-valueasdate-using-local
